homeassistant/components/isy994.py

```diff
@@ -44,190 +43,352 @@
 
-    for component in SUPPORTED_DOMAINS:
+
+def _categorize_programs(hass: HomeAssistant, programs: dict) -> None:
+    """Categorize the ISY994 programs."""
+    for domain in SUPPORTED_PROGRAM_DOMAINS:
         try:
-            folder = ISY.programs[KEY_MY_PROGRAMS]['HA.' + component]
+            folder = programs[KEY_MY_PROGRAMS]['HA.{}'.format(domain)]
         except KeyError:
             pass
         else:
             for dtype, _, node_id in folder.children:
-                if dtype is KEY_FOLDER:
-                    program = folder[node_id]
+                if dtype == KEY_FOLDER:
+                    entity_folder = folder[node_id]
                     try:
-                        node = program[KEY_STATUS].leaf
-                        assert node.dtype == 'program', 'Not a program'
-                    except (KeyError, AssertionError):
-                        pass
-                    else:
-                        if component not in PROGRAMS:
-                            PROGRAMS[component] = []
-                        PROGRAMS[component].append(program)
-
-
-def _categorize_weather() -> None:
+                        status = entity_folder[KEY_STATUS]
+                        assert status.dtype == 'program', 'Not a program'
+                        if domain != 'binary_sensor':
+                            actions = entity_folder[KEY_ACTIONS]
+                            assert actions.dtype == 'program', 'Not a program'
+                        else:
+                            actions = None
+                    except (AttributeError, KeyError, AssertionError):
+                        _LOGGER.warning("Program entity '%s' not loaded due "
+                                        "to invalid folder structure.",
+                                        entity_folder.name)
+                        continue
+
+                    entity = (entity_folder.name, status, actions)
+                    hass.data[ISY994_PROGRAMS][domain].append(entity)
+
+
+def _categorize_weather(hass: HomeAssistant, climate) -> None:
     """Categorize the ISY994 weather data."""
-    global WEATHER_NODES
-
-    climate_attrs = dir(ISY.climate)
-    WEATHER_NODES = [WeatherNode(getattr(ISY.climate, attr), attr,
-                                 getattr(ISY.climate, attr + '_units'))
+    climate_attrs = dir(climate)
+    _LOGGER.debug(climate)
+    _LOGGER.debug(climate_attrs)
+    weather_nodes = [WeatherNode(getattr(climate, attr),
+                                 attr.replace('_', ' '),
+                                 getattr(climate, '{}_units'.format(attr)))
                      for attr in climate_attrs
-                     if attr + '_units' in climate_attrs]
+                     if '{}_units'.format(attr) in climate_attrs]
+    hass.data[ISY994_WEATHER].extend(weather_nodes)
 
 
 def setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the ISY 994 platform."""
+    if ISY994_NODES not in hass.data:
```

Comment:
always true

Reply:
Doesn't setup get called multiple times if multiple `isy994` blocks are specified in the config? I took this paradigm from here and I assumed that's why it was there: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/dyson.py#L47-L48

If I've got that wrong I can remove the check

Reply:
The bootstrap setup a component only one time but the platform it could be call multiple times.